lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -154,31 +181,67 @@ bool ArrayAllocation::mapInitializationStores() {
 /// allocated.
 bool ArrayAllocation::isInitializationWithKnownElements() {
   ArraySemanticsCall Uninitialized(Alloc, "array.uninitialized");
-  if (Uninitialized &&
-      (ArrayValue = Uninitialized.getArrayValue()) &&
-      (ElementBuffer = Uninitialized.getArrayElementStoragePointer()))
+  if (Uninitialized && (ArrayValue = Uninitialized.getArrayValue()) &&
+      (ElementBuffer = Uninitialized.getArrayElementStoragePointer())) {
+    Substitutions =
+        ArrayValue->getType().gatherAllSubstitutions(Alloc->getModule());
     return mapInitializationStores();
+  }
 
   return false;
 }
 
+bool ArrayAllocation::replacementsAreValid() {
+  auto ElementCount = ElementValueMap.size();
+  llvm::SmallVector<SILValue, 4> ElementValueVector;
```

Comment:
Why not sink this below the early exit. The compiler may do it, but it may not.

Also, where are you using this below, isn't it dead?